### ISSUE TITLE
Error in loss calculation when using ignore_index

### ISSUE THREAD

**Issue by idc9** [qubvel/segmentation_models.pytorch#852]
Thanks for the very nice segmentation package! 

In the current implementation of `SoftBCEWithLogitsLoss`, if you use `ignore_index` then the loss is not calculated correctly if `reduction=mean`. In particular, the current implementation multiplies each sample's loss by 1 or 0 depending on if the sample should be included; when you calculate the mean you are including the samples that should be ignored in the denominator (i.e. the loss would be too small). See [this line](https://github.com/qubvel/segmentation_models.pytorch/blob/6db76a1106426ac5b55f39fba68168f3bccae7f8/segmentation_models_pytorch/losses/soft_bce.py#L74). Similar behavior seems to occur in `SoftBCEWithLogitsLoss`, `label_smoothed_nll_loss`, and `DiceLoss`.

I assume the goal is to mirror the behavior of standard pytorch e.g. see the documentation [CrossEntropyLoss](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html) or you can check

```python
import torch.nn.functional as F
import torch
import numpy as np

# 3 inputs, mean reduciton
y_pred = torch.from_numpy(np.array([[.5, .5, .5], [0, 0, 0]]).T)
y_true = torch.from_numpy(np.array([1, 1, 1])).reshape(3,)
F.cross_entropy(input=y_pred, target=y_true, reduction='mean')
```

```
tensor(0.9741, dtype=torch.float64)
```
```python
# 3 inputs, ignore 1, mean reduction
y_pred = torch.from_numpy(np.array([[.5, .5, .5], [0, 0, 0]]).T)
y_true = torch.from_numpy(np.array([1, 1, -100])).reshape(3,)
F.cross_entropy(input=y_pred, target=y_true, ignore_index=-100, reduction='mean')
# we expect this number to be be the same as above
```

```
tensor(0.9741, dtype=torch.float64)
```
